hs/WebParsing/TimeTableParser.hs

```diff
@@ -0,0 +1,280 @@
+creates a tutorial from the information currently residing in the CourseSlot
+----------------------------------------------------------------------------------------}
+makeTutorial :: CourseSlot -> Maybe Tutorial -> Tutorial
+makeTutorial slot Nothing =
+  Tutorial {  tutorialSection = Just (slotSection slot),
+              times = [],
+              timeStr = (slotTime_str slot)
+            }
+makeTutorial slot (Just tutorial) =
+  tutorial {timeStr = T.append (timeStr tutorial) (slotTime_str slot)}
+
+{----------------------------------------------------------------------------------------
+makes a lecture from the information currently residing in the courseSlot
+----------------------------------------------------------------------------------------}
+makeLecture :: CourseSlot -> Maybe Lecture -> Lecture
+makeLecture slot Nothing = 
+ Lecture { extra = (slotExtra slot),
+              section = (slotSection slot),
+              cap = (slotCap slot),
+              time_str = (slotTime_str slot),
+              time = (slotTime slot),
+              instructor = (slotInstructor slot),
+              enrol = (slotEnrol slot),
+              wait = (slotWait slot)
+            }
+makeLecture slot (Just lec) = lec {time_str = T.append (time_str lec) (slotTime_str slot)}
+
+{----------------------------------------------------------------------------------------
+updates the courseSlot record with information from a given row
+----------------------------------------------------------------------------------------}
+updateSlot :: [T.Text] -> CourseSlot -> CourseSlot
+updateSlot row slot =
+  let partialUpdate = slot {
+            slotTime_str = T.takeWhile (/= ' ') (row !! 5),
+            slotInstructor = (row !! 7)}
+  in  if (row !! 3) == "" --then no new section
+      then partialUpdate
+      else partialUpdate {slotSection = (row !! 3)}
+
+{----------------------------------------------------------------------------------------
+Takes in a 2-d list of 'cells' corresponding to a single course. Each lecture and tutori-
+al can span more than a single row if it has seperate timeslots. 
+
+we recurse over each 'row'. We check to see if the 4th cell contains the name of a lectu-
+re or tutorial section, in which case we consolidate the current lecture or tutorial into
+the session, and create a new lecture of tutorial.
+
+there are five cases, the last of which is useless and only in place for pattern complet-
+eness. 
+  1. no more information, with last block being a tutorial. 
+  2. no more information, with last block being a lecture.
+  3. moer information with last block being a lecture
+  4. more information with last block being a tutorial.
+within 3. and 4, if the current row is 
+----------------------------------------------------------------------------------------}
+updateSession :: [[T.Text]] -> CourseSlot -> Maybe Tutorial -> Maybe Lecture -> Session -> Session
```

Comment:
Look into the `Either` type here instead of using two parameters for each of a tutorial and lecture